drivers/pktmem.c

```diff
@@ -140,14 +156,67 @@ static ssize_t net_allocPktBuf(void **bufp)
 
 	SYS_ARCH_UNPROTECT(old_level);
 
-	*bufp = (void *)ph + CACHE_LINE_SIZE + i * PKT_BUF_SIZE;
+	*bufp = (void *)ph - pkt_buf_opts.page_head_offset + (i * (pkt_buf_opts.alloc_sz / pkt_buf_opts.pkt_cnt));
+
+	return pkt_buf_opts.pkt_sz;
+}
+
 
-	return PKT_BUF_SIZE;
+static unsigned net_whichAlwaysZero(const void *p)
+{
+	return 0;
+}
+
+
+static unsigned net_whichFromAddr(const void *p)
+{
+	size_t next_pow2_pkt_cnt = (pkt_buf_opts.pkt_cnt <= 2) ? pkt_buf_opts.pkt_cnt : ((1ull << 32) >> __builtin_clz(pkt_buf_opts.pkt_cnt - 1));
+	size_t pkt_buf_idx = __builtin_ctz(PAGE_SIZE) - __builtin_ctz(next_pow2_pkt_cnt); /* log2(PAGE_SIZE) - ceil(log2(pkt_buf_opts.pkt_cnt)) */
+	return (((size_t)(p) & (PAGE_SIZE - 1)) >> pkt_buf_idx);
+}
+
+
+int net_pktMemInit(size_t pkt_buf_max_sz)
+{
+	if (net_isPktMemInitialized()) {
+		return 0;
+	}
+
+	if (pkt_buf_max_sz == 0) {
+		return -EINVAL;
+	}
+
+	size_t page_head_sz;
+
+	pkt_buf_opts.alloc_sz = PAGE_SIZE * ((pkt_buf_max_sz / PAGE_SIZE) + 1);
+
+	if (pkt_buf_max_sz + CACHE_LINE_SIZE <= pkt_buf_opts.alloc_sz) {
+		/* Cache line optimisation */
+		page_head_sz = CACHE_LINE_SIZE;
+		pkt_buf_opts.pkt_cnt = PAGE_SIZE / (pkt_buf_max_sz + page_head_sz);
+		pkt_buf_opts.which = net_whichFromAddr;
+	}
+	else {
+		/* No memory - skip the optimisation */
```

Comment:
This should never happen, DMA can now overwrite `buf_page_head_t` header.

Reply:
`pkt_buf_max_sz + CACHE_LINE_SIZE > pkt_buf_opts.alloc_sz`

Reply:
`pkt_buf_opts.alloc_sz` should be large enough to hold at least one packet buffer (`pkt_buf_max_sz`) and the header (`CACHE_LINE_SIZE`).